src/Microsoft.AspNet.Mvc.TagHelpers/ScriptTagHelper.cs

```diff
@@ -51,83 +125,161 @@ public class ScriptTagHelper : TagHelper
         protected internal ILogger<ScriptTagHelper> Logger { get; set; }
 
         [Activate]
+        protected internal IHostingEnvironment HostingEnvironment { get; set; }
+
+        [Activate]
         protected internal ViewContext ViewContext { get; set; }
 
+        [Activate]
+        protected internal IMemoryCache Cache { get; set; }
+
+        // Internal for ease of use when testing.
+        protected internal GlobbingUrlBuilder GlobbingUrlBuilder { get; set; }
+
         /// <inheritdoc />
         public override async Task ProcessAsync(TagHelperContext context, TagHelperOutput output)
         {
-            if (!AttributeMatcher.AllRequiredAttributesArePresent(context, ViewContext, RequiredAttributes, Logger))
-            {
-                if (Logger.IsEnabled(LogLevel.Verbose))
-                {
-                    Logger.WriteVerbose(
-                        "Skipping processing for {0} with ID {1} on view '{2}'",
-                        nameof(ScriptTagHelper),
-                        context.UniqueId,
-                        ViewContext.View.Path);
-                }
+            var modeResult = AttributeMatcher.DetermineMode(context, ModeDetails);
+
+            Debug.Assert(modeResult.FullMatches.Select(match => match.Mode).Distinct().Count() <= 1,
+                $"There should only be one mode match, check the {nameof(ModeDetails)}");
 
+            modeResult.LogDetails(Logger, this, context.UniqueId, ViewContext.View.Path);
+
+            if (!modeResult.FullMatches.Any())
+            {
+                // No attributes matched so we have nothing to do
                 return;
             }
 
-            var content = new StringBuilder();
+            var mode = modeResult.FullMatches.First().Mode;
 
-            // NOTE: Values in Output.Attributes are already HtmlEncoded
+            // NOTE: Values in TagHelperOutput.Attributes are already HtmlEncoded
+            var attributes = new Dictionary<string, string>(output.Attributes);
+            
+            var builder = new StringBuilder();
+            var originalContent = await context.GetChildContentAsync();
 
-            // We've taken over rendering here so prevent the element rendering the outer tag
-            output.TagName = null;
+            if (mode == Mode.Fallback && string.IsNullOrEmpty(SrcInclude))
+            {
+                // No globbing to do, just build a <script /> tag to match the original one in the source file
+                BuildScriptTag(attributes, originalContent, builder);
+            }
+            else
+            {
+                BuildGlobbedScriptTags(originalContent, attributes, builder);
```

Comment:
could the `BuildScriptTag` and `BuildGlobbedScriptTag` methods have the same signature?  odd to switch two parameters between them